Client/Frontend/Login Management/LoginListViewController.swift

```diff
@@ -166,6 +176,23 @@ class LoginListViewController: UIViewController {
     }
 }
 
+// MARK: - Favicon Loading
+extension LoginListViewController {
+
+    private func loadFaviconsForLogins(logins: [Login]) -> Deferred<Maybe<[GUID: [Favicon]]>> {
+        let urls = logins.map { $0.hostname }
+        return self.profile.favicons.getFaviconsForHistoryURLs(urls) >>== { favicons in
+            let filteredFavicons = favicons.asArray().flatMap { $0 }
+            var loginFaviconMap = [GUID: [Favicon]]()
+            logins.forEach { login in
+                // Filter by login hostname -> history URL and extract out the favicon
+                loginFaviconMap[login.guid] = filteredFavicons.filter { $0.0.hasPrefix(login.hostname) } .map { $0.1 } ?? []
```

Comment:
This is the worst kind of line-painter problem: for each login you're walking a potentially long list to find the ones that match. This is O(n^2): n logins, walking a list of n or more favicons with the filter.

You want this to be O(n) + O(log n) — so make sure the list of returned favicons is attributed by hostname if you can, then partition them using something like `SequenceType.groupBy`, then walk the logins doing a lookup in the hash table.